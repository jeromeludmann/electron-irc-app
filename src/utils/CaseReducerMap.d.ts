import { Action, Reducer } from 'redux'

export interface CaseReducerMap<R> {
  [action: string]: R
}
