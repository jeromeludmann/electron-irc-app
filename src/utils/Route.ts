import { Action } from 'redux'
import { CHANNEL_REGEXP } from '@app/utils/helpers'

export interface Route {
  serverKey: string
  bufferKey: string
}

export interface RoutedAction<T = string> extends Action<T> {
  route: Route
}

export const BufferKey = {
  STATUS: '@status',
  RAW: '@raw',
  ALL: '@all',
  ACTIVE: '@active',
  NONE: '@none',
}
export const isRaw = (key: string) => {
  return key === BufferKey.RAW
}

export const isStatus = (key: string) => {
  return key === BufferKey.STATUS
}

export const isChannel = (key: string) => {
  return CHANNEL_REGEXP.test(key)
}

export const isPrivate = (key: string) => {
  return !isRaw(key) && !isStatus(key) && !isChannel(key)
}
