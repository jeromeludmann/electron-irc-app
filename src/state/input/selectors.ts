import { createSelector } from 'reselect'
import { getBufferFactory } from '@app/state/buffer/selectors'
import { Route } from '@app/utils/Route'

export function getInput(route?: Route) {
  return createSelector(getBufferFactory(route), channel => channel.input)
}

export function getInputValue(route?: Route) {
  return createSelector(getInput(route), input => input.value)
}
