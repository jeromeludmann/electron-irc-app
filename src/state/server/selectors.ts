import { createSelector } from 'reselect'
import { getServers } from '@app/state/root/selectors'
import { getRoute } from '@app/state/route/selectors'

export function getServerFactory(serverKey?: string) {
  return createSelector(
    getServers,
    getRoute,
    (servers, route) => servers[serverKey || route.serverKey],
  )
}

export function getBuffers(serverKey?: string) {
  return createSelector(getServerFactory(serverKey), server => server.buffers)
}
