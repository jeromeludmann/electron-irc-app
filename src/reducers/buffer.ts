import { createSelector } from 'reselect'
import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from '@app/reducers/messages'
import { InputState, inputInitialState, reduceInput } from '@app/reducers/input'
import { RoutedAction } from '@app/utils/Route'
import { RouteState, selectRoute } from '@app/reducers/route'
import { RECEIVE_JOIN, ReceiveJoinAction } from '@app/actions/msgIncoming'
import { SWITCH_WINDOW } from '@app/actions/ui'
import { ServerState, selectBuffers } from '@app/reducers/server'
import { CaseReducerMap } from '@app/utils/CaseReducerMap'

type BufferPartialState = Readonly<{
  activity: boolean
}>

export type BufferState = Readonly<{
  input: InputState
  messages: MessagesState
}> &
  BufferPartialState

type BufferReducer<S = BufferState> = (
  buffer: S,
  action: RoutedAction,
  extraStates: {
    route: RouteState
    server: ServerState
  },
) => S

export const bufferInitialState = {
  activity: false,
  input: inputInitialState,
  messages: messagesInitialState,
}

const caseReducers: CaseReducerMap<BufferReducer<BufferPartialState>> = {
  [RECEIVE_JOIN]: (_, action: ReceiveJoinAction, extraStates) => ({
    activity: action.payload.user.nick !== extraStates.server.user.nick,
  }),

  [SWITCH_WINDOW]: () => ({ activity: false }),
}

export const reduceBuffer: BufferReducer = (
  buffer = bufferInitialState,
  action,
  extraStates,
): BufferState => ({
  ...(action.type in caseReducers
    ? caseReducers[action.type](buffer, action, extraStates)
    : buffer),
  input: reduceInput(buffer.input, action),
  messages: reduceMessages(buffer.messages, action, extraStates),
})

export const selectBuffer = createSelector(
  selectBuffers,
  selectRoute,
  (buffers, route) => buffers[route.bufferKey],
)

export const selectActivity = createSelector(
  selectBuffer,
  buffer => buffer.activity,
)
