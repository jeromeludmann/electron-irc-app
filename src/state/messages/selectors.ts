import { createSelector } from 'reselect'
import { getBufferFactory } from '@app/state/buffer/selectors'
import { Route } from '@app/utils/Route'

export function getMessages(route?: Route) {
  return createSelector(getBufferFactory(route), buffer => buffer.messages)
}
