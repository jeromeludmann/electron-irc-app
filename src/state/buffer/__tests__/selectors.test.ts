import { getBufferFactory } from '@app/state/buffer/selectors'
import { rootInitialState } from '@app/state/root/reducer'
import { Route } from '@app/utils/Route'

describe('getBufferFactory()', () => {
  it('should get "getBuffer" selector', () => {
    expect(getBufferFactory()(rootInitialState)).toMatchSnapshot()
  })
})

describe('getBufferFactory(route)', () => {
  it('should get "getBuffer" selector', () => {
    const route: Route = { serverKey: 'serverKey', bufferKey: '#channel' }
    expect(getBufferFactory(route)(rootInitialState)).toMatchSnapshot()
  })
})
