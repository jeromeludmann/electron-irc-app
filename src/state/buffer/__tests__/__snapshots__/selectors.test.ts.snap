// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBufferFactory() should get "getBuffer" selector 1`] = `
Object {
  "activity": false,
  "input": Object {
    "dirtyValue": "",
    "history": Object {
      "index": 0,
      "values": Array [],
    },
    "value": "",
  },
  "messages": Array [],
}
`;

exports[`getBufferFactory(route) should get "getBuffer" selector 1`] = `undefined`;
